Stock management regression fixes

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

 Stock managment fixes:

  1. Use `$already_reduced_stock` instead of also considering `$refunded_item_quantity` while deleting orders. This will bring back part of #27504 again, but for now this seems to be the best solution for countering #28605. It needs discussion whether deleting a line item completely should also undo any refund related changes on it or not.

  2. Also mark `stock_reduced` flag on order if any of the line item has any `_reduced_stock` flag. This will allow for stock restoring logic to work properly when order is cancelled.

  3. Only adjust line item stock when order is in `processing`, `completed` or `on-hold` status state, because we need to reduce stocks on these status only. Stock adjustments in refunds or when changing statuses is already taken care of by their specific hooks.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #28605 #28505 #28585  . Reopen #27504 for more consideration

### How to test the changes in this Pull Request:

1. Follow steps in #28605 #28505 and #28585

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Adjust stock items only for statuses which reduces the stock.

vedanshujain committed:
Add unit tests for all status change ops.

This loops and tests stock changes through ALL supported WooCommerce statuses. Hopefully this will make our inventory tracking more robust.

vedanshujain committed:
Stock managment fixes:

1. Use $already_reduced_stock instead of also considering $refunded_item_quantity while deleting orders. This will bring back part of #27504 again, but for now this seems to be the best solution for countering #28605. It needs discussion whether deleting a line item completely should also undo any refund related changes on it or not.

2. Also mark `stock_reduced` flag on order if any of the line item has any `_reduced_stock` flag. This will allow for stock restoring logic to work properly when order is cancelled.

3. Only adjust line item stock when order is in `processing`, `completed` or `on-hold` status state, because we need to reduce stocks on these status only. Stock adjustments in refunds or when changing statuses is already taken care of by their specific hooks.

roykho reviewed (dismissed):
With these new changes, are the older unit tests still needed? https://github.com/woocommerce/woocommerce/blob/master/tests/php/includes/admin/class-wc-admin-functions-test.php#L72-L288

Otherwise your PR tested well.

peterfabian reviewed (dismissed):
Hi, mostly it seems to be working fine, I just found a couple of issues/questions:

I'm not sure if this is in scope, but these cases seems to not work for me as I would expect: 
- from wp admin, adding a product to an order does not reduce the stock (and any subsequent changes also do nothing to the stock level for me)
- reinstantiating an order with refunds

Complete list of what I tested:
1. create order from the front end---ok
2. reduce stock from wp admin by edit button---ok
3. increase stock from wp admin by the edit button---ok
4. refund some stock---ok
5. delete line item---ok
6. add the line item back---stock level not ok
6. new order from front end---ok
7. refund 2 items---ok
7. Cancel the order---ok
7. back to On hold---not ok (it decreases based on the full stock amount, not the refunded amount)
8. editing qty of item in the order---ok (fixes the overall count to correct count, i.e. fixes the problem in the previous step)
9. removing line item---ok

Maybe we should update the notice when deleting an item from an order, too?
![Screenshot 2020-12-18 at 14 41 34](https://user-images.githubusercontent.com/2207451/102621001-3bfb4480-413f-11eb-9506-5561e63cde75.png)

vedanshujain committed:
Merge branch 'master' into fix/28505

vedanshujain committed:
Better docs for test functions to specify expectations.

vedanshujain commented on tests/php/includes/wc-stock-functions-tests.php:
```diff
@@ -0,0 +1,185 @@
+	public function test_status_transition_stock_reduce_to_stock_reduce() {
+		foreach ( $this->order_stock_reduce_statuses as $order_status_from ) {
+			foreach ( $this->order_stock_reduce_statuses as $order_status_to ) {
+				$this->transition_order_status_and_assert_stock_quantity( $order_status_from, $order_status_to, 9, 9 );
```
Looks like adding it in assert messages is a bit confusing because assert can fail whenever expected and actual stock level would not match, which can happen for a lot of reason than just stock being reduced twice (for example, there could be a bug which increases the stock, and assert message will say stock is not reduced twice).

But I agree that its a bit unclear at the moment, so I have added expectations in doc comments for all these tests in https://github.com/woocommerce/woocommerce/pull/28620/commits/5171108967750ad8c6fe7ab87ccd4d6139268578. Let me know if these makes sense.

vedanshujain commented on tests/php/includes/wc-stock-functions-tests.php:
```diff
@@ -0,0 +1,185 @@
+		$this->assertEquals( $before_transition, $product->get_stock_quantity() );
+
+		// Changing status from UI also calls this method in metadata save hook. This has impact on stock levels, so simulate it here as well.
+		wc_save_order_items( $order, $order->get_items() );
```
Yeah, I missed this initially but only noticed it when tests were passing but there was clearly a bug while doing the same steps via interface. Adding this made the tests fail when I was expecting them to, so I agree that code around these parts is bit fragile, and is a good candidate for refactoring efforts.

vedanshujain commented on tests/php/includes/wc-stock-functions-tests.php:
```diff
@@ -0,0 +1,185 @@
+	/**
+	 * Helper function to simulate creating order from cart.
+	 *
+	 * @param string $status_from Status for the newly created order.
```
Addressed in https://github.com/woocommerce/woocommerce/pull/28620/commits/5171108967750ad8c6fe7ab87ccd4d6139268578

vedanshujain commented:
@peterfabian 
> from wp admin, adding a product to an order does not reduce the stock (and any subsequent changes also do nothing to the stock level for me)

This is expected, we removed stock changes unless status is also changed to prevent ghost stock issues. See https://github.com/woocommerce/woocommerce/pull/26642.

> reinstantiating an order with refunds

This is also kinda expected, because as part of this PR I am reverting some of the related changes to have more discussion. See from the PR description: 

> Use $already_reduced_stock instead of also considering $refunded_item_quantity while deleting orders. This will bring back part of #27504 again, but for now this seems to be the best solution for countering #28605. It needs discussion whether deleting a line item completely should also undo any refund related changes on it or not.

claudiosanches committed:
Fixed coding standards

claudiosanches reviewed (approved):
Seems to be working as expected, I followed all steps to reproduce and I agree with your comments in https://github.com/woocommerce/woocommerce/pull/28620#issuecomment-748907594

I'm not merging now because it's still missing unit test to run on Travis (in localhost those works fine) after I fixed some coding standards.

vedanshujain commented:
Merging because manually tested the failed e2e tests (able to delete refund). there is another one about setup wizard failing, but it does not looks related.
